### ISSUE TITLE
Get session cookies when using --cookies-from-browser

### ISSUE THREAD

**Issue by flashdagger** [yt-dlp/yt-dlp#5534]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

While investigating #4514 I found a different behavior between `--cookies cookies.txt` and `--cookies-from-browser chrome`.
The first option injected all cookies into the request, while the latter one was missing the session cookies.

Session cookies have `expires=None` and `discard=True`. Taking a look into the codebase I see that session cookies imported from Chrome have `expires=0` and `discard=False`.

https://github.com/yt-dlp/yt-dlp/blob/d7b460d0e5fc710950582baed2e3fc616ed98a80/yt_dlp/cookies.py#L335-L339

Is this intended or possibly a bug?

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
NA
```

**Comment by pukkandan** [comment]
Definitely not intended. I thought chrome doesn't write session cookies to disk 🤔. If it does, we should "fix" the cookies similar to https://github.com/yt-dlp/yt-dlp/blob/b96187a9603aaa7ead8ad94ac82433f085ea4895/yt_dlp/utils.py#L1654-L1667

**Comment by flashdagger** [comment]
>  I thought chrome doesn't write session cookies to disk 🤔

It does because of the `Continue where you left off` startup option.

**Comment by Tblue** [comment]
Firefox is also affected by this issue: It seems like session cookies are not stored in its `cookies.sqlite` file.

Instead, they are in `sessionstore.jsonlz4` (but that file only exists if Firefox is not running). That's a file in Mozilla's custom `mozLz4` format, which essentially is just the LZ4 block format preceded by a lower-endian 32-bit unsigned int specifying the size of the uncompressed data in bytes. 

Once uncompressed, the `sessionstore` data can be parsed as JSON, and the session cookies found in the top-level key `cookies` of the resulting object.

**Comment by StefanLobbenmeier** [comment]
Was about to open the same issue, here is what I found:

Disclaimer at the start - I did this investigation with a recent version of brave, will need to check if this applies to all chromium browsers and if so from which version.

The brave cookies database has 2 columns related to cookie expiry: expires_utc and has_expires. yt-dlp only checks the expires_utc column:

- For session cookies expires_utc is set to 0 and has_expires to 0. 
- For expiring cookies expires_utc is set to a utc timestamp and has_expires to 1

The current logic in yt-dlp takes expires_utc only and the phyton cooki will consider a value of 0 as expired:
```python
class Cookie:
    # ...
    def is_expired(self, now=None):
        if now is None: now = time.time()
        if (self.expires is not None) and (self.expires <= now):
            return True
        return False
```

To support session_cookies we should therefore set expires to None, since they do not expire.

**Comment by StefanLobbenmeier** [comment]
@Tblue can you elaborate on what you mean by removing the header? I could confirm that this file is created when I close firefox:


It starts like this:

```
mozLz40≈¡{"version":["sesÚ
restore",1],"windows":[{"tab	bentrie
```

I tried removing the `mozLz40` or the `mozLz40≈¡` using a text editor (probably not the right tool 😄) but passing that to lz4 I always get `Error 44 : Unrecognized header : file cannot be decoded`, what do I have to remove exactly?

**Comment by StefanLobbenmeier** [comment]
Actually nevermind, I found this superuser thread on it: https://superuser.com/a/1363751

And it linked this python gist that I could test sucessfully: https://gist.github.com/Tblue/62ff47bef7f894e92ed5

**Comment by Tblue** [comment]
Yup, and I was going to link to that Gist - I’ve written it. 🙂

**Comment by Tblue** [comment]
Sure, feel free to use the Gist under the UNLICENSE.

**Comment by Tblue** [comment]
@StefanLobbenmeier, you might also find this script useful as an inspiration (feel free to treat it as licensed under the UNLICENSE as well): [dump-firefox-cookies.py](https://gist.github.com/Tblue/8c6ee7c29b683dcf3e5dad2ff1edc338)

**Comment by bashonly** [comment]
Keeping open until this is solved for Firefox as well

**Comment by StefanLobbenmeier** [comment]
Yeah did not have the time to implement for Firefox, but it’s still on my radar. If someone else wants to take it I don’t mind though :)

**Comment by Fd929c2CE5fA** [comment]
While investigating the issue, I found @StefanLobbenmeier's PR #9747 and this issue.


> Instead, they are in `sessionstore.jsonlz4` (but that file only exists if Firefox is not running). That's a file in Mozilla's custom `mozLz4` format, which essentially is just the LZ4 block format preceded by a lower-endian 32-bit unsigned int specifying the size of the uncompressed data in bytes.

Yes, `sessionstore.jsonlz4` is only generated after the browser is closed and disappears while the browser is running.

But don't forget we also have the `sessionstore-backups` folder, which contains different backups:

- `previous.jsonlz4`: This is the snapshot of the session before the current one. It is used to restore the previous session state if needed.

- `recovery.jsonlz4`: This file represents the current ongoing session state, saved periodically during browsing. It is used for restoring the current session when Firefox unexpectedly closes.

- `recovery.baklz4`: This is a backup of recovery.jsonlz4, offering an additional fallback if the primary recovery file is corrupted.

- `upgrade.jsonlz4-{timestamp}`: These files are older session backups created during Firefox upgrades. They capture the session state at the time of upgrade and can be used to restore sessions from that point.

I believe `recovery.jsonlz4` should be the preferred one, as it is constantly updated while the browser is running.
